src/plugins/dataflow-tool/dataflow-logger.ts

```diff
@@ -0,0 +1,88 @@
+import { LogEventName, Logger, DataflowProgramChange } from "../../lib/logger";
+import { Connection, Control, Node } from "rete";
+
+type DataflowLogPayload = any; //Node | Connection | Control | object;
+
+export function dataflowLogEvent( operation: string, payload: DataflowLogPayload, tileId: string ){
+  const logEventName = LogEventName.DATAFLOW_TOOL_CHANGE;
+
+  if (payload instanceof Node){
+    const n = payload;
+    const change: DataflowProgramChange = {
+      targetType: 'node',
+      nodeTypes: [n.name],
+      nodeIds: [n.id]
+    };
+    Logger.logToolChange(logEventName, operation, change, tileId);
+  }
+
+  else if (payload instanceof Connection){
+    const outputNode = payload.output.node as Node;
+    const inputNode = payload.input.node as Node;
+
+    const change: DataflowProgramChange = {
+      targetType: 'connection',
+      nodeTypes: [outputNode.name, inputNode.name],
+      nodeIds: [outputNode.id, inputNode.id],
+      connectionOutputNodeId: outputNode.id,
+      connectionOutputNodeType: outputNode.name,
+      connectionInputNodeId: inputNode.id,
+      connectionInputNodeType: inputNode.name
+    };
+    Logger.logToolChange(logEventName, operation, change, tileId);
+  }
+
+  else if (payload instanceof Control){
+    const ctrl = payload as Control;
+    const node = payload.parent as Node;
+
+    if (ctrl && node){
+      const change: DataflowProgramChange = {
+        targetType: 'nodedropdown',
+        nodeTypes: [node.name],
+        nodeIds: [node.id],
+        selectItem: ctrl.key,
+        value: (ctrl as any).props.value,
+        units: (ctrl as any).props.currentUnits || ""
+      };
+      Logger.logToolChange(logEventName, operation, change, tileId);
+    }
+  }
+
+  // we pass a plain object for title change and sensor selections
+  // so we differentiate by keys sent in object
+  else {
+    const changeProperties = Object.keys(payload);
+
+    if (changeProperties.includes("programTitleValue")){
+      const change: DataflowProgramChange = {
+        targetType: 'program',
+        programTitle: payload.programTitleValue
+      };
+      Logger.logToolChange(logEventName, operation, change, tileId);
+    }
+
+    if (changeProperties.includes("sensorTypeValue")){
```

Comment:
I notice these are all `if`s (not `else if`s), so we could hit all of them. Is that desired behavior?

Reply:
Since `changeProperties` is an array, I thought I might end up populating it with various combinations of things in order to assemble the change object - but that was all in the abstract and I see now each of them are mutually exclusive so I've changed them to `else if`.  Thanks for pointing that out.